### ISSUE TITLE
removing sample(s) from the talon database

### ISSUE THREAD

**Issue by ew367** [mortazavilab/TALON#104]
Hi

How does talon process datasets that have only partially been added to the input database?  When running talon again using the same config file and input databse will it continue from part way through the partial dataset, or skip the dataset since part of it is already present? My HPC job was interrupted during processing and I want to know how I can tell whether the sample that was being added at the time has been completed, or if data is still missing after running talon again.

Thanks

**Comment by callumparr** [comment]
Also interested to know. When this happened I just deleted the database and initialize a new one as there isn’t a —resume flag.

**Comment by ew367** [comment]
> Also interested to know. When this happened I just deleted the database and initialize a new one as there isn’t a —resume flag.

Yes, that has been my approach in the past too, but my new dataset is HUGE and had already been running for nearly 2 weeks before the interruption so I really don't want to do that this time!

**Comment by fairliereese** [comment]
You can use this python code to check if the dataset has been added to your database:

```python
import sqlite3
import pandas as pd

db = 'database_name.db'
with sqlite3.connect(db) as conn:
     q = 'SELECT dataset_name FROM dataset'
     datasets = pd.read_sql_query(q, conn)

print(datasets.dataset_name.tolist())
```

TALON is pretty good about discarding or not pushing incomplete changes to the database but this is not a surefire method. What I typically do is I make a backup copy of my TALON database before trying to add new datasets to it. That way, if the run fails, I can simply restart using the backup. I'm sorry there's not a better way to do this but this is definitely something that I learned based on getting burned in the past as well.

**Comment by callumparr** [comment]
> What I typically do is I make a backup copy of my TALON database before trying to add new datasets to it. That way, if the run fails, I can simply restart using the backup

That's simple and ingenious. Not sure why I did not think to do that. TY for replies as always.

**Comment by rb520826** [comment]
Hi, thanks for the above suggestions. I am also interested in the removal of a sample from the database - was there any update to whether this is possible please?

Thanks!

**Comment by callumparr** [comment]
> You can use this python code to check if the dataset has been added to your database:
> 
> ```python
> import sqlite3
> import pandas as pd
> 
> db = 'database_name.db'
> with sqlite3.connect(db) as conn:
>      q = 'SELECT dataset_name FROM dataset'
>      datasets = pd.read_sql_query(q, conn)
> 
> print(datasets.dataset_name.tolist())
> ```
> 
> TALON is pretty good about discarding or not pushing incomplete changes to the database but this is not a surefire method. What I typically do is I make a backup copy of my TALON database before trying to add new datasets to it. That way, if the run fails, I can simply restart using the backup. I'm sorry there's not a better way to do this but this is definitely something that I learned based on getting burned in the past as well.

Is it possible to extract the sample description, the second column in the config file. 

I been playing around with sqlite3 module trying to get the column headers of the dataset table in the database but its a beyond me.

**Comment by fairliereese** [comment]
> Is it possible to extract the sample description, the second column in the config file.
>
>I been playing around with sqlite3 module trying to get the column headers of the dataset table in the database but its a beyond me.

You should be able to pull that info out using the following sql query: `SELECT DISTINCT sample FROM dataset`

As an aside, if you're interested in navigating the stuff in the TALON database, I'd definitely recommend downloading a DB viewer such as [this one](https://sqlitebrowser.org/). You can look through the tables and write / test out queries on your tables so you don't have to open up python and sqlite3 every time you want to poke around. 

As another aside, I am much more comfortable in pandas in python than I am in manipulating these tables through sqlite3. If that's more your speed, there are sqlite3 functions that will literally dump a table from a database into a pandas table (see [here](https://datacarpentry.org/python-ecology-lesson/09-working-with-sql/index.html#accessing-data-stored-in-sqlite-using-python-and-pandas) for example) to make it easy to work on.
